### ISSUE TITLE
Isssue with custom dataset transfer learning on Alexnet

### ISSUE THREAD

**Issue by SokPhanith** [wang-xinyu/tensorrtx#560]
## Env

- GPU, jetson Nano.
- Jetpak 4.5.1
- Cuda 10.2
- TensorRT 7.1

## About this repo

I train my own data-set  4 classes with Transfer Learning on Alexnet , Just change last layer output out_features=4 and gen wts after than I rebuild on folder Alexnet and edit  file alex.cpp with line : 
 
IFullyConnectedLayer* fc3 = network->addFullyConnected(*relu7->getOutput(0),num_classes, weightMap["classifier.6.weight"], weightMap["classifier.6.bias"]);  

num_classes = 4

## problem

after then I build engine got issue : 

Loading weights: ../alexnet.wts
set name out
[05/22/2021-16:33:46] [E] [TRT] (Unnamed Layer* 17) [Fully Connected]: bias weights has count 1000 but 4 was expected
[05/22/2021-16:33:46] [E] [TRT] Could not compute dimensions for prob, because the network is not valid.
[05/22/2021-16:33:46] [E] [TRT] Network validation failed.
build out
alexnet: /home/phanith/tensorrtx/alexnet/alex.cpp:174: void APIToModel(unsigned int, nvinfer1::IHostMemory**): Assertion `engine != nullptr' failed.
Aborted (core dumped)

### My code training : 
import numpy as np
import torch
from torch import nn
from torch import optim
import torch.nn.functional as F
from torchvision import datasets, transforms, models
import struct
import os
train_dir = 'singel_board/train/'
valid_dir = 'singel_board/valid/'
batch_size = 8
num_classes = 4
epochs = 30
print_every = 200
img_size = 224
feature_extract = True
model_name = 'alex'
output_model = 'output'
torch.manual_seed(4321)
train_transforms = transforms.Compose([transforms.Resize((img_size,img_size)),
                                       transforms.RandomHorizontalFlip(),
                                       transforms.RandomRotation(4),
                                       transforms.RandomCrop(224),
                                       transforms.RandomVerticalFlip(p=0.5),
                                       transforms.RandomAutocontrast(p=0.5),
                                       transforms.ToTensor(),
                                       transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
                                      ])
valid_transforms = transforms.Compose([transforms.Resize((img_size,img_size)),
                                       transforms.ToTensor(),
                                       transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])])
valid_data = datasets.ImageFolder(valid_dir,transform=train_transforms)
train_data = datasets.ImageFolder(train_dir,transform=valid_transforms)

trainloader = torch.utils.data.DataLoader(train_data, batch_size=batch_size,num_workers=2,shuffle=True)
validloader = torch.utils.data.DataLoader(valid_data, batch_size=batch_size,num_workers=2,shuffle=False)

classes = trainloader.dataset.classes
print('classes : ',classes)
print('Number Dataloader for train : ',len(trainloader))
print('Number Dataloader for valid : ',len(validloader))
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print('Device for training : ',device)

def set_parameter_requires_grad(model, feature_extracting):
    if feature_extracting:
        for param in model.parameters():
            param.requires_grad = False
net = models.alexnet(pretrained=True)
set_parameter_requires_grad(net, feature_extracting=feature_extract)
num_ftrs = net.classifier[6].in_features
net.classifier[6] = nn.Linear(num_ftrs,num_classes)
net = net.to(device)
print(net)

criterion = nn.CrossEntropyLoss()
params_to_update = net.parameters()
if feature_extract:
    params_to_update = []
    for name,param in net.named_parameters():
        if param.requires_grad == True:
            params_to_update.append(param)
            print("\t",name)
else:
    for name,param in net.named_parameters():
        if param.requires_grad == True:
            print("\t",name)
optimizer = optim.SGD(params_to_update, lr=0.001, momentum=0.9)
for epoch in range(epochs):  
    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        inputs, labels = data[0].to(device), data[1].to(device)
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        if i % print_every == (print_every-1): 
            net.eval()
            test_loss = 0
            total_time = 0
            true = 0
            with torch.no_grad():
                for inputs, labels in validloader:
                    inputs, labels = inputs.to(device),labels.to(device)
                    logps = net(inputs)
                    _, predicted = torch.max(logps.data, 1)
                    batch_loss = criterion(logps, labels)
                    test_loss += batch_loss.item()
                    true += (predicted == labels).sum().item()
                    total_time += labels.size(0)
            print('........................................')
            print('training       : [%d, %5d] loss: %.3f' %(epoch + 1, i + 1, running_loss / print_every))
            print('validtion      : [%d, %5d] loss: %.3f' %(epoch + 1, i + 1, test_loss / total_time))
            print('Accuracy valid : %d %%' % (100 * true / total_time))
            #torch.save(net.state_dict() ,output_model+"/"+model_name+"_"+str((100 * true / total_time))+".pth")
            running_loss = 0.0
            net.train()
print('Finished Training')
net.eval()
correct = 0
total = 0
with torch.no_grad():
    for data in validloader:
        images, labels = data[0].to(device),data[1].to(device)
        outputs = net(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
print('Accuracy of the network on the validtion images: %d %%' % (100 * correct / total))

correct_pred = {classname: 0 for classname in classes}
total_pred = {classname: 0 for classname in classes}
with torch.no_grad():
    for data in validloader:
        images, labels = data[0].to(device),data[1].to(device)
        outputs = net(images)
        _, predictions = torch.max(outputs, 1)
        for label, prediction in zip(labels, predictions):
            if label == prediction:
                correct_pred[classes[label]] += 1
            total_pred[classes[label]] += 1

for classname, correct_count in correct_pred.items():
    accuracy = 100 * float(correct_count) / total_pred[classname]
    print("Accuracy for class {:5s} is: {:.1f} %".format(classname,accuracy))

print('Gen wts file for TensorRT Engine')
f = open("alexnet.wts", 'w')
f.write("{}\n".format(len(net.state_dict().keys())))
for k,v in net.state_dict().items():
    print('key: ', k)
    print('value: ', v.shape)
    vr = v.reshape(-1).cpu().numpy()
    f.write("{} {}".format(k, len(vr)))
    for vv in vr:
        f.write(" ")
        f.write(struct.pack(">f", float(vv)).hex())
    f.write("\n")
tmp = torch.ones(1,3,224,224).to(device)
outputs = net(tmp)
print(outputs)
torch.save(net,"alexnet.pth")

**Comment by wang-xinyu** [comment]
```
[05/22/2021-16:33:46] [E] [TRT] (Unnamed Layer* 17) [Fully Connected]: bias weights has count 1000 but 4 was expected
```
means your .wts is not compatible with the network definition. You need to double check this fc layer.

**Comment by SokPhanith** [comment]
I did it. I must edit 2 place in file alex.cpp 
 - IFullyConnectedLayer* fc3 = network->addFullyConnected(*relu7->getOutput(0),4, weightMap["classifier.6.weight"], 
     weightMap["classifier.6.bias"]);
 - OUTPUT_SIZE = 4
And one more thing I would to ask you @wang-xinyu , when I read in python TensorRT engine fps around 42-43 and just a simple test on GPU jetson nano 34-35 fps . The engine will build out is Float32 or Float16.

**Comment by aditya-dl** [comment]
@SokPhanith it's set to float32 by default during the network generation phase, you can test around with float16. However, I think you will need to generate .wts in float16 as well.

**Comment by SokPhanith** [comment]
@aditya-dl  Thank for reply. I really want test with float16 on Jetson nano but something like Alexnet I don't know where to edit file alex.cpp and a new python to generate *.wts to build TensorRT engine float16. plz give me some example

**Comment by aditya-dl** [comment]
@SokPhanith Try changing the Dataype in this line: https://github.com/wang-xinyu/tensorrtx/blob/d9e2cbc294ae61bcd6b877156756cb06c057e9b1/alexnet/alex.cpp#L173 

You can replace it with kHALF as mentioned here: https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/namespacenvinfer1.html#a83aed11a1c160f30dcd13809678bdd29

I haven't tried generating the *.wts with float16, however, the TensorRT developer guide mentions that if specified float16 while building the engine, it should automatically take care of everything.

**Comment by wang-xinyu** [comment]
.wts should always be fp32, the trt APIs can load fp32 weights, and convert it to fp16 when building the engine.

**Comment by wang-xinyu** [comment]
closing it.
